### PR TITLE
Enable minimal datagram support

### PR DESCRIPTION
**Pull request by mangolas**
## Description

Adds the missing (minimal) support for sending and receiving datagrams. 

## Testing

Tests do not cover datagrams.

## Documentation

Documentation in general in Rust side is very minimal.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -373,3 +373,5 @@ target/
 
 # Cargo (Rust) lock file
 Cargo.lock
+.idea
+cmake-build-debug/
```
Comment (nibanks) [inline .gitignore:377]: Do we need these?

File: scripts/build.rs [hunk 2]
```diff
@@ -11,12 +12,26 @@ fn main() {
...
+
+    match target.as_str() {
+        "x86_64-apple-darwin" => config
+            .define("CMAKE_OSX_ARCHITECTURES", "x86_64")
+            .define("CMAKE_OSX_DEPLOYMENT_TARGET", "10.15"),
+        "aarch64-apple-darwin" => config
+            .define("CMAKE_OSX_ARCHITECTURES", "arm64")
+            .define("CMAKE_OSX_DEPLOYMENT_TARGET", "11.0"),
+        _ => &mut config
+    };
```
Comment (nibanks) [inline scripts/build.rs:32]: Is this for macOS support?

File: src/lib.rs [hunk 3]
```diff
@@ -1207,6 +1221,11 @@ impl Settings {
         self.idle_timeout_ms = value;
         self
     }
+    pub fn set_datagram_receive_enabled(&mut self, value: bool) -> &mut Settings {
+        self.is_set_flags |= (value as u64) << 27;
```
Comment (nibanks) [inline src/lib.rs:1225]: This we really be using `value` here? You are setting the receive enabled flag (to true or false) so I think this should always be set.

File: src/lib.rs [hunk 4]
```diff
@@ -1310,7 +1329,7 @@ impl Drop for Registration {
 impl Configuration {
     pub fn new(
         registration: &Registration,
-        alpn: &Buffer,
+        alpn: &[Buffer],
```
Comment (nibanks) [inline src/lib.rs:1332]: Why the changes in this section?

File: .gitignore (hunk 1 above)
Comment (mangolas) [reply .gitignore:377]: Not necessarily, wouldn't hurt either, but can be removed.

File: scripts/build.rs (hunk 2 above)
Comment (mangolas) [reply scripts/build.rs:32]: Yes, this allows to run eg. 'cargo build --target x86_64-apple-darwin` and so, in near future there is a need for Android targets as well.

File: .gitignore (hunk 1 above)
Comment (nibanks) [reply .gitignore:377]: We can leave it. Thanks.

File: scripts/build.rs (hunk 2 above)
Comment (nibanks) [reply scripts/build.rs:32]: Ok. Thanks. Perhaps it's work updating the Cargo GitHub action to test macOS to validate?

File: src/lib.rs (hunk 4 above)
Comment (mangolas) [reply src/lib.rs:1332]: Configuration::new Rust function was missing a way to actually give more than one alpn parameter, so it had to be changed anyways, so I'm proposing here using Rust slice rather than reference and separate length. Listener `start` is another place alpn is passed so I changed that as well to be uniform.

File: src/lib.rs (hunk 3 above)
Comment (nibanks) [reply src/lib.rs:1225]: For instance, look at the functions above, such as `set_idle_timeout_ms`. The `is_set_flags` is always set because you are modifying the actual setting value (independent of it being true or false). So you should always be setting the appropriate bit in `is_set_flags` here. For instance:
```suggestion
        self.is_set_flags |= 1 << 27;
```

File: src/lib.rs (hunk 4 above)
Comment (nibanks) [reply src/lib.rs:1332]: Ok.

File: src/lib.rs (hunk 3 above)
Comment (mangolas) [reply src/lib.rs:1225]: Aah, finally got it, sure you are right, will change.

File: scripts/build.rs (hunk 2 above)
Comment (nibanks) [reply scripts/build.rs:32]: In cargo.yml try adding `macos-latest` to `os: [ubuntu-latest, windows-latest]`?

Comment (mangolas) [reply scripts/build.rs:32]: Done.

**Review by nibanks — APPROVED** [review]

**Comment by nibanks** [comment]
Thanks for the change!
